### PR TITLE
Fix mempool generatePopData

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: src/pop/blockchain/mempool_block_tree.cpp [hunk 1]
```diff
@@ -271,6 +271,36 @@ void applyPayloadsOrRemoveIfInvalid(
...
+  const auto& maxSize = tree_->getParams().getMaxPopDataSize();
+
+  auto popSize = pop.estimateSize();
+
+  std::vector<VTB>& vtbs = pop.vtbs;
+  for (auto it = vtbs.rbegin(); it != vtbs.rend();) {
+    it = popSize > maxSize ? (popSize -= it->estimateSize(),
+                              std::vector<VTB>::reverse_iterator{
+                                  vtbs.erase(std::next(it).base())})
+                           : (std::next(it));
```
Comment (blackyblack) [inline src/pop/blockchain/mempool_block_tree.cpp:284]: Make  normal `if` statement.

File: src/pop/blockchain/mempool_block_tree.cpp [hunk 2]
```diff
@@ -271,6 +271,36 @@ void applyPayloadsOrRemoveIfInvalid(
...
 }
 
+void MemPoolBlockTree::filterPopDataBySize(PopData& pop) {
+  const auto& maxSize = tree_->getParams().getMaxPopDataSize();
+
+  auto popSize = pop.estimateSize();
+
+  std::vector<VTB>& vtbs = pop.vtbs;
+  for (auto it = vtbs.rbegin(); it != vtbs.rend();) {
+    it = popSize > maxSize ? (popSize -= it->estimateSize(),
```
Comment (blackyblack) [inline src/pop/blockchain/mempool_block_tree.cpp:281]: popSize is unsigned. May overflow here.

File: src/pop/blockchain/mempool_block_tree.cpp [hunk 3]
```diff
@@ -271,6 +271,36 @@ void applyPayloadsOrRemoveIfInvalid(
...
+                              std::vector<VTB>::reverse_iterator{
+                                  vtbs.erase(std::next(it).base())})
```
Comment (blackyblack) [inline src/pop/blockchain/mempool_block_tree.cpp:283]: What's going on here? Please rewrite or add some meaningful comment.

File: src/pop/blockchain/mempool_block_tree.cpp [hunk 4]
```diff
@@ -315,6 +345,9 @@ void MemPoolBlockTree::filterInvalidPayloads(PopData& pop) {
   applyPayloadsOrRemoveIfInvalid(mutator, pop.vtbs, continueOnInvalidContext);
   applyPayloadsOrRemoveIfInvalid(mutator, pop.atvs, continueOnInvalidContext);
 
+  // filter payloads by size
+  filterPopDataBySize(pop);
```
Comment (Warchant) [inline src/pop/blockchain/mempool_block_tree.cpp:349]: I don't like this. I would like to see size validation during setState.

File: src/pop/blockchain/mempool_block_tree.cpp [hunk 5]
```diff
@@ -271,6 +271,40 @@ void applyPayloadsOrRemoveIfInvalid(
...
+  std::vector<VTB>& vtbs = pop.vtbs;
+  for (auto it = vtbs.begin(); it != vtbs.end(); ++it) {
+    if (popSize + it->estimateSize() >= maxSize) {
+      vtbs.erase(it, vtbs.end());
+      break;
+    }
+    popSize += it->estimateSize();
+  }
+
+  std::vector<ATV>& atvs = pop.atvs;
```
Comment (Warchant) [inline src/pop/blockchain/mempool_block_tree.cpp:289]: popSize = pop.estimateSize();

if(popSize <= maxSize) return;

File: src/pop/blockchain/mempool_block_tree.cpp [hunk 6]
```diff
@@ -271,6 +271,40 @@ void applyPayloadsOrRemoveIfInvalid(
...
 
+void MemPoolBlockTree::filterPopDataBySize(PopData& pop) {
+  const auto& maxSize = tree_->getParams().getMaxPopDataSize();
+
+  // initialize with the empty pop data size
+  auto popSize = PopData{}.estimateSize();
+
+  std::vector<VTB>& vtbs = pop.vtbs;
+  for (auto it = vtbs.begin(); it != vtbs.end(); ++it) {
+    if (popSize + it->estimateSize() >= maxSize) {
```
Comment (Warchant) [inline src/pop/blockchain/mempool_block_tree.cpp:282]:
```suggestion
    if (popSize + it->estimateSize() > maxSize) {
```

File: src/pop/blockchain/mempool_block_tree.cpp [hunk 7]
```diff
@@ -271,6 +271,40 @@ void applyPayloadsOrRemoveIfInvalid(
...
+      vtbs.erase(it, vtbs.end());
+      break;
+    }
+    popSize += it->estimateSize();
+  }
+
+  std::vector<ATV>& atvs = pop.atvs;
+  for (auto it = atvs.begin(); it != atvs.end(); ++it) {
+    if (popSize + it->estimateSize() >= maxSize) {
```
Comment (Warchant) [inline src/pop/blockchain/mempool_block_tree.cpp:291]:
```suggestion
    if (popSize + it->estimateSize() > maxSize) {
```

File: src/pop/blockchain/mempool_block_tree.cpp [hunk 8]
```diff
@@ -271,6 +271,40 @@ void applyPayloadsOrRemoveIfInvalid(
...
+      atvs.erase(it, atvs.end());
+      break;
+    }
+    popSize += it->estimateSize();
+  }
+
+  std::vector<VbkBlock>& vbks = pop.context;
+  for (auto it = vbks.begin(); it != vbks.end(); ++it) {
+    if (popSize + it->estimateSize() >= maxSize) {
```
Comment (Warchant) [inline src/pop/blockchain/mempool_block_tree.cpp:300]:
```suggestion
    if (popSize + it->estimateSize() > maxSize) {
```

File: src/pop/blockchain/mempool_block_tree.cpp [hunk 9]
```diff
@@ -271,6 +271,40 @@ void applyPayloadsOrRemoveIfInvalid(
   payloads.erase(it, payloads.end());
 }
 
+void MemPoolBlockTree::filterPopDataBySize(PopData& pop) {
+  const auto& maxSize = tree_->getParams().getMaxPopDataSize();
+
+  // initialize with the empty pop data size
+  auto popSize = PopData{}.estimateSize();
```
Comment (Warchant) [inline src/pop/blockchain/mempool_block_tree.cpp:278]:
```suggestion
  auto popSize = PopData{}.estimateSize() + 20;
```

File: src/pop/blockchain/mempool_block_tree.cpp [hunk 10]
```diff
@@ -271,6 +271,40 @@ void applyPayloadsOrRemoveIfInvalid(
...
+
+  std::vector<ATV>& atvs = pop.atvs;
+  for (auto it = atvs.begin(); it != atvs.end(); ++it) {
+    if (popSize + it->estimateSize() >= maxSize) {
+      atvs.erase(it, atvs.end());
+      break;
+    }
+    popSize += it->estimateSize();
+  }
+
```
Comment (Warchant) [inline src/pop/blockchain/mempool_block_tree.cpp:297]: popSize = pop.estimateSize();

if(popSize <= maxSize) return;

File: src/pop/blockchain/mempool_block_tree.cpp [hunk 11]
```diff
@@ -271,6 +271,40 @@ void applyPayloadsOrRemoveIfInvalid(
...
+  std::vector<VbkBlock>& vbks = pop.context;
+  for (auto it = vbks.begin(); it != vbks.end(); ++it) {
+    if (popSize + it->estimateSize() >= maxSize) {
+      vbks.erase(it, vbks.end());
+      break;
+    }
+    popSize += it->estimateSize();
+  }
+}
```
Comment (Warchant) [inline src/pop/blockchain/mempool_block_tree.cpp:306]: add assert that pop.estimateSize <= maxSize

**Comment by Warchant** [comment]
```
if(pop.estimateSize() <= maxSize) return;
```

File: include/veriblock/pop/blockchain/pop/continue_on_invalid_context.hpp [hunk 12]
```diff
@@ -19,37 +19,63 @@ struct ContinueOnInvalidContext {
...
 
+    vbks_size += vbk_size;
     vbks++;
     return true;
   }
 
  private:
-  size_t vtbs = 0;
+  bool canFitSize(size_t size) {
+    size_t popdatasize = PopData{}.estimateSize() + singleBEValueSize(atvs) +
```
Comment (Warchant) [inline include/veriblock/pop/blockchain/pop/continue_on_invalid_context.hpp:66]:
```suggestion
    size_t popdatasize = sizeof(PopData::version) + singleBEValueSize(atvs) +
```

**Review by Warchant — APPROVED** [review]

**Pull request merged**